pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,520 @@
+            b_num = np.array([[p1[0], p1[2]], [p3[0], p3[2]]])
+        else:
+            a_num = np.array([[p3[0], p3[1]], [p2[0], p2[1]]])
+            b_num = np.array([[p1[0], p1[1]], [p3[0], p3[1]]])
+        coeffs_0 = np.linalg.det(a_num) / denom
+        coeffs_1 = np.linalg.det(b_num) / denom
+        coeffs.append(coeffs_0)
+        coeffs.append(coeffs_1)
+        return coeffs
+
+    def get_positions(self, structure: Structure, points: list) -> Dict[Tuple[int, int, int], list]:
+        """
+        Calculates all the positions of each hkl point in the 2D diffraction pattern. Distance in centimeters.
+        Args:
+            structure (Structure): The input structure.
+            points (list): All points to be checked.
+        Returns:
+            dict of hkl plane to xy-coordinates.
+        """
+        positions = {}
+        points = self.zone_axis_filter(points)
+        # first is the max_d, min_r
+        first_point_dict = self.get_first_point(structure, points)
+        for point in first_point_dict:
+            first_point = point
+            first_d = first_point_dict[point]
+        spacings = self.get_interplanar_spacings(structure, points)
+        # second is the first non-parallel-to-first-point vector when sorted.
+        # note 000 is "parallel" to every plane vector.
+        for plane in sorted(spacings.keys()):
+            second_point, second_d = plane, spacings[plane]
+            if not self.is_parallel(first_point, second_point):
+                break
+        p1 = first_point
+        p2 = second_point
+        if (0, 0, 0) in points:
+            points.remove((0, 0, 0))
+        points.remove(first_point)
+        points.remove(second_point)
+        positions[(0, 0, 0)] = np.array([0, 0])
+        r1 = 10 ** 10 * self.wavelength_rel() * self.camera_length / first_d
+        positions[first_point] = np.array([r1, 0])
+        r2 = 10 ** 10 * self.wavelength_rel() * self.camera_length / second_d
+        phi = np.arccos(
+            np.dot(p1, np.transpose(p2)) / (np.sqrt(np.dot(p1, np.transpose(p1)) * np.dot(p2, np.transpose(p2)))))
+        positions[second_point] = np.array([r2 * np.cos(phi), r2 * np.sin(phi)])
+        denom = np.linalg.det(np.array([[p1[0], p1[1]], [p2[0], p2[1]]]))
+        init_denom_0 = (denom == 0)
+        if init_denom_0:
+            denom = np.linalg.det(np.array([[p1[0], p1[2]], [p2[0], p2[2]]]))
+        for plane in points:
+            coeffs = self.get_plot_coeffs(p1, p2, plane, denom, init_denom_0)
+            pos = np.array([coeffs[0] * positions[first_point][0] + coeffs[1] * positions[second_point][0],
+                            coeffs[0] * positions[first_point][1] + coeffs[1] * positions[second_point][1]])
+            positions[plane] = pos
+        points.append((0, 0, 0))
+        points.append(first_point)
+        points.append(second_point)
+
+        return positions
+
+    def TEM_dots(self, structure: Structure, points: list) -> list:
```

Comment:
capitalization in function name, PEP8 violation

Reply:
Thanks!